### ISSUE TITLE
fluid.clip.GradientClipByNorm does not work as expected

### ISSUE THREAD

**Issue by gavin1332** [PaddlePaddle/Paddle#9368]
To test the gradient clipping, I make some experiments on the GRU-RNN Language Model:
* The origin model works fine
* Add the gradient_clip attribute (GradientClipByNorm) to the embedding layer, and it triggers "nan" ppl (maybe, gradient_clip does not support sparse updating)
* Add the gradient_clip attribute (GradientClipByNorm) to the last FC op (dense updating), with a common max_norm=5.0, and the "nan" ppl is triggered off again
* Update max_norm=1e6, which I think is large enough to avoid gradient clipping, but it still triggers off "nan"

the code is here
``` python
import os
import sys 
import math

import numpy as np

import paddle.v2 as paddle
import paddle.fluid as fluid

# set hyper parameters
N = 1000

epoch_num = 40
batch_size = 20

emb_size = 650 
hid_size = 650 

cuda_id = 1 
base_lr = 1.0 
emb_lr_x = 1.0 
l1_lr_x = 1.0 
fc_lr_x = 1.0 
dropout_rate = 0.5 
init_bound = 0.05

model_dir = "model/gru_clip"
os.system("mkdir -p " + model_dir)
os.system("cp %s %s/" % (sys.argv[0], model_dir))

vocab = paddle.dataset.imikolov.build_dict(0)
vocab_size = len(vocab)
print "vocab_size: %d" % vocab_size

# build computational graph
src_wordseq = fluid.layers.data(name="src_wordseq", shape=[1], dtype="int64", lod_level=1)
dst_wordseq = fluid.layers.data(name="dst_wordseq", shape=[1], dtype="int64", lod_level=1)

emb = fluid.layers.embedding(input=src_wordseq, size=[vocab_size, emb_size],
        param_attr=fluid.ParamAttr(initializer=fluid.initializer.Uniform(low=-init_bound, high=init_bound),
            #gradient_clip=fluid.clip.GradientClipByNorm(clip_norm=100000.0), 
            learning_rate=emb_lr_x),
        is_sparse=True)

emb = fluid.layers.dropout(x=emb, dropout_prob=dropout_rate, is_test=False)

fc0 = fluid.layers.fc(input=emb, size=hid_size*3,
        param_attr=fluid.ParamAttr(initializer=fluid.initializer.Uniform(low=-init_bound, high=init_bound), learning_rate=l1_lr_x))
gru_h0 = fluid.layers.dynamic_gru(input=fc0, size=hid_size,
        param_attr=fluid.ParamAttr(initializer=fluid.initializer.Uniform(low=-init_bound, high=init_bound), learning_rate=l1_lr_x))

gru_h0 = fluid.layers.dropout(x=gru_h0, dropout_prob=dropout_rate, is_test=False)

fc = fluid.layers.fc(input=gru_h0, size=vocab_size, act='softmax',
        param_attr=fluid.ParamAttr(initializer=fluid.initializer.Uniform(low=-init_bound, high=init_bound),
            gradient_clip=fluid.clip.GradientClipByNorm(1000000.0),
            learning_rate=fc_lr_x))

cost = fluid.layers.cross_entropy(input=fc, label=dst_wordseq)
average_cost = fluid.layers.mean(x=cost)

sgd_optimizer = fluid.optimizer.SGD(
        learning_rate=fluid.layers.exponential_decay(
            learning_rate=base_lr,
            decay_steps=2100*6,
            decay_rate=0.8,
            staircase=True))
sgd_optimizer.minimize(average_cost)

# run train graph
def to_lodtensor(data, place):
    seq_lens = [len(seq) for seq in data]
    cur_len = 0
    lod = [cur_len]
    for l in seq_lens:
        cur_len += l
        lod.append(cur_len)
    flattened_data = np.concatenate(data, axis=0).astype("int64")
    flattened_data = flattened_data.reshape([len(flattened_data), 1])
    res = fluid.core.LoDTensor()
    res.set(flattened_data, place)
    res.set_lod([lod])
    return res

place = fluid.CUDAPlace(cuda_id)
exe = fluid.Executor(place)

train_reader = paddle.batch(
        paddle.reader.shuffle(paddle.dataset.imikolov.train(vocab, N, data_type=paddle.dataset.imikolov.DataType.SEQ), buf_size=N),
        batch_size)

exe.run(fluid.default_startup_program())
for epoch_id in xrange(epoch_num):
    print "epoch_%d start" % epoch_id

    i = 0
    for data in train_reader():
        i += 1
        lod_src_wordseq = to_lodtensor(map(lambda x: x[0], data), place)
        lod_dst_wordseq = to_lodtensor(map(lambda x: x[1], data), place)
        ret_average_cost = exe.run(
                fluid.default_main_program(),
                feed={"src_wordseq": lod_src_wordseq, "dst_wordseq": lod_dst_wordseq},
                fetch_list=[average_cost])
        average_ppl = math.exp(ret_average_cost[0])
        if i % 100 == 0:
            print "step %d ppl: %.3f" % (i, average_ppl)
    print "total steps:", i

    # save model
    save_dir = "%s/epoch_%d" % (model_dir, epoch_id)
    feed_var_names = ["src_wordseq", "dst_wordseq"]
    fetch_vars = [fc, average_cost]
    fluid.io.save_inference_model(save_dir, feed_var_names, fetch_vars, exe)

    print "epoch_%d finished ." % epoch_id
```

**Comment by pkuyym** [comment]
Thanks for your feedback. Seems weird, let me reproduce first.
